stratum/hal/lib/barefoot/bfrt_packetio_manager.cc

```diff
@@ -271,36 +250,38 @@ ::util::Status BfrtPacketioManager::TransmitPacket(
   std::string buf;
   RETURN_IF_ERROR(DeparsePacketOut(packet, &buf));
 
-  bf_pkt* pkt = nullptr;
-  bf_pkt_tx_ring_t tx_ring = BF_PKT_TX_RING_0;
-  RETURN_IF_BFRT_ERROR(
-      bf_pkt_alloc(device_id_, &pkt, buf.size(), BF_DMA_CPU_PKT_TRANSMIT_0));
-  auto pkt_cleaner =
-      gtl::MakeCleanup([pkt, this]() { bf_pkt_free(device_id_, pkt); });
-  RETURN_IF_BFRT_ERROR(bf_pkt_data_copy(
-      pkt, reinterpret_cast<const uint8_t*>(buf.data()), buf.size()));
-  RETURN_IF_BFRT_ERROR(bf_pkt_tx(device_id_, pkt, tx_ring, pkt));
-  pkt_cleaner.release();
+  RETURN_IF_ERROR(bf_sde_interface_->TxPacket(device_, buf));
 
   return ::util::OkStatus();
 }
 
-::util::Status BfrtPacketioManager::HandlePacketRx(bf_dev_id_t dev_id,
-                                                   bf_pkt* pkt,
-                                                   bf_pkt_rx_ring_t rx_ring) {
+// TODO: drop Sde in name?
+::util::Status BfrtPacketioManager::HandleSdePacketRx() {
+  std::unique_ptr<ChannelReader<std::string>> reader;
   {
     absl::ReaderMutexLock l(&data_lock_);
     if (!initialized_) RETURN_ERROR(ERR_NOT_INITIALIZED) << "Not initialized.";
+    reader = ChannelReader<std::string>::Create(packet_receive_channel_);
   }
-  ::p4::v1::PacketIn packet;
-  std::string buffer(reinterpret_cast<const char*>(bf_pkt_get_pkt_data(pkt)),
-                     bf_pkt_get_pkt_size(pkt));
-  RETURN_IF_ERROR(ParsePacketIn(buffer, &packet));
-  {
-    absl::WriterMutexLock l(&rx_writer_lock_);
-    rx_writer_->Write(packet);
+
+  while (true) {
+    std::string buffer;
+    int code = reader->Read(&buffer, absl::InfiniteDuration()).error_code();
+    if (code == ERR_CANCELLED) break;
+    if (code == ERR_ENTRY_NOT_FOUND) {
+      LOG(ERROR) << "Read with infinite timeout failed with ENTRY_NOT_FOUND.";
+      continue;
+    }
+
+    ::p4::v1::PacketIn packet_in;
+    RETURN_IF_ERROR(ParsePacketIn(buffer, &packet_in));
+    {
+      LOG(WARNING) << "Getting rx_writer_lock_ " << &rx_writer_lock_;
```

Comment:
Note: use VLOG for this logging before merging the PR

Reply:
I think this was for debugging and can be removed now.